### PR TITLE
Chip optimizations

### PR DESCRIPTION
**Pull request by jerryz123**
* Support single-porting accumulator through the use of accumulator "sub-banks"
* Support clock-gating Gemmini modules
* Support sharing SPAD/ACC between Int8 and FP gemminis
* Reduce bitwidths of loop unroller multipliers and adders
* Fix error where small portion of scratchpad was unusable when double-buffering in the loop unrollers

When single-porting the accumulator banks, input-dilated convs will sometimes fail because they keep writing to the same accumulator banks. A different write pattern will have to be found eventually for those cases, but that's outside the scope of this PR.

### CODE REVIEW

File: src/main/scala/gemmini/Scratchpad.scala [hunk 1]
```diff
@@ -444,9 +445,8 @@ class Scratchpad[T <: Data, U <: Data, V <: Data](config: GemminiArrayConfig[T,
         ex_read_resp.valid := bio.read.resp.valid && !bio.read.resp.bits.fromDMA
         ex_read_resp.bits := bio.read.resp.bits
 
-        val dma_read_pipe = Pipeline(Queue(dma_read_resp), mem_pipeline)
-        val ex_read_pipe = Pipeline(Queue(ex_read_resp), mem_pipeline)
```
Comment (hngenc) [inline src/main/scala/gemmini/Scratchpad.scala]: @jerryz123 Are these queues necessary for functional correctness? Or were they just added to make physical design easier?

Comment (jerryz123) [reply src/main/scala/gemmini/Scratchpad.scala]: These don't seem to be in the latest commit on this branch? I think they were not necessary.

Comment (hngenc) [reply src/main/scala/gemmini/Scratchpad.scala]: I removed them myself, but I just wanted to make sure that they weren't necessary

**Comment by hngenc** [comment]
We now pass CI for the `defaultConfig`, but we still fail CI for the dual-chip-config. I'm trying to figure out why

File: src/main/scala/gemmini/AccumulatorMem.scala [hunk 2]
```diff
@@ -222,17 +309,15 @@ class AccumulatorMem[T <: Data, U <: Data](
   val q_will_be_empty = (q.io.count +& q.io.enq.fire()) - q.io.deq.fire() === 0.U
   io.read.req.ready := q_will_be_empty && (
       // Make sure we aren't accumulating, which would take over both ports
-      !(io.write.fire() && io.write.bits.acc) &&
-      // Make sure we aren't reading something that is still being written
-      !(RegNext(io.write.fire()) && RegNext(io.write.bits.addr) === io.read.req.bits.addr) &&
-      !(w_buf_valid && waddr_buf === io.read.req.bits.addr) &&
+      !io.write.valid &&
```
Comment (hngenc) [inline src/main/scala/gemmini/AccumulatorMem.scala:312]: @jerryz123 Why did we remove the `&& io.write.acc` check here?

File: src/main/scala/gemmini/AccumulatorMem.scala [hunk 3]
```diff
@@ -44,17 +44,59 @@ class AccumulatorWriteReq[T <: Data: Arithmetic](n: Int, t: Vec[Vec[T]]) extends
...
 
-  override def cloneType: this.type = new AccumulatorMemIO(n, t, scale_t).asInstanceOf[this.type]
+  val ext_mem = if (use_shared_ext_mem) Some(Vec(acc_sub_banks, new ExtMemIO)) else None
+
+  val acc = new Bundle {
+    val valid = Output(Bool())
+    val ina = Output(t.cloneType)
+    val inb = Output(t.cloneType)
+    val out = Input(t.cloneType)
+  }
```
Comment (hngenc) [inline src/main/scala/gemmini/AccumulatorMem.scala:61]: What does this bundle mean exactly? What does `ina`, `inb`, etc. mean?

**Pull request merged**